### ISSUE TITLE
cmd/go: `go mod tidy` ignores `go.work` file

### ISSUE THREAD

**Issue by bozaro** [golang/go#50750]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.18beta1 linux/amd64
</pre>

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/bozaro/.cache/go-build"
GOENV="/home/bozaro/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/bozaro/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/bozaro/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/lib/go-1.18"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/lib/go-1.18/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.18beta1"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/bozaro/github/go-work-play/tools/go.mod"
GOWORK="/home/bozaro/github/go-work-play/go.work"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build4292218461=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

### What did you do?

Minimal reproducing repository: https://github.com/bozaro/go-work-play/tree/go-mod-tidy

Full script:
```
$ git clone https://github.com/bozaro/go-work-play.git -b go-mod-tidy
$ cd go-work-play/tools
$ go mod tidy
go: finding module for package github.com/bozaro/go/go-work-play/shared/foo
github.com/bozaro/go/go-work-play/tools/hello imports
	github.com/bozaro/go/go-work-play/shared/foo: module github.com/bozaro/go@latest found (v0.0.0-20200925035954-2333c6299f34), but does not contain package github.com/bozaro/go/go-work-play/shared/foo
```

### What did you expect to see?
I expect `go.mod` and `go.sum` updated with current working copy state.

### What did you see instead?
I see that `go mod tidy` try to get modules for `shared` go modules from repository ignoring `go.work` content.

**Comment by bcmills** [comment]
Tidiness is a property of an individual module, not a workspace: if a module is tidy, then a downstream consumer of the module knows which versions to use for every dependency of every package in that module.

If you don't particularly care about downstream consumers having a package that is provided by the workspace, you can use `go mod tidy -e` to ignore the error from the missing package.

Otherwise, you either need to publish the workspace dependencies before running `go mod tidy` (so that they have well-defined upstream versions), or tell the `go` command what those versions will be (using a `replace` and `require` directive in the individual module).

**Comment by bcmills** [comment]
@matloob, for Go 1.19 I wonder if we should augment the `go.work` file to be able to declare explicit intended versions for the modules in the workspace. Then `go mod tidy` could use those versions instead of looking upstream. 🤔

**Comment by bozaro** [comment]
Currently with reproducing repository (https://github.com/bozaro/go-work-play/tree/go-mod-tidy).

I can run:
```
cd tools/hello
go run .
```
And `go run` works like workspace go modules declared as `replace` in current `tools/go.mod`.

But for `go mod tidy` I need add `replace` manually.

This behaviour looks like inconsistent: I expects that `go run` and `go mod tidy` would use same dependencies.

**Comment by liadmord** [comment]
What valid use cases are there for `go.work` if we are still required to fully implement the `go.mod` in order to work on multi module repo?

**Comment by AlmogBaku** [comment]
This is also happening for `go get ./...`

I agree with @liadmord, the only benefit I see get from the `go.work` is the better "understanding" of my IDE
Or - maybe I'm doing something wrong here?

**Comment by Jictyvoo** [comment]
I agree with @bozaro @liadmord and @AlmogBaku 

I have a monorepo private project where I have multiple modules, and on each one I need to insert a lot of `replace` statements for each direct and indirect dependencies. With `go workspaces` I really want to be possible to run `go mod download` or `go mod tidy` without the needed to write a **replace** statement in every `go.mod` file in my workspace.

**Comment by springeye** [comment]
Many of my colleagues have this need

**Comment by jincheng9** [comment]
a big problem for go 1.18 workspace mode

**Comment by maxsxu** [comment]
Encountering the same issue with `go1.18.2`.

Furthermore, some other `go mod` commands also failed.

- `go mod vendor`

```
➜  server   go mod vendor 
atmax.io/go-workspace/server imports
        atmax.io/go-workspace/commons/pkg: no required module provides package atmax.io/go-workspace/commons/pkg; to add it:
        go get atmax.io/go-workspace/commons/pkg
```

- `go mod verify `

```
➜  server   go mod verify 
atmax.io/go-workspace/server : missing ziphash: open hash: no such file or directory
```

**Comment by Gbps** [comment]
Doing a `go mod tidy` on a `go.work` enabled workspace should just treat the module as if you added a dependency directly to the git repo url, but the repo is on your filesystem instead.

For example, if I have a dependency `github.com/Gbps/mymodule` and I specify its location in a `go.work`, it would make sense to me that `go mod tidy` would treat the dependency as if it went to `github.com` and pulled the latest version, but pulled it from my local git repo instead. 

That by far makes the most sense to me. Once it's in my `go.mod`, then I can push the local repo to remote and anyone who uses my `go.mod` after will get the same dependency that used to be on my local machine.

Having to edit `go.mod` at all to use a local `go.work` really feels like it defeats the purpose of the workspace addition.

**Comment by opengs** [comment]
This is a very big issue for me in the case of creating Docker containers :( I cant use `go mod tidy` and copy the dependency graph to the Dockerfile. So every time I build a container I have to download downstream dependencies instead of doing `go mod download`. 

Golang is considered the language for microservices and In the current state of the `go work` it's very hard to work with them in monorepo. Creating multiple-container environment is incredibly hard :(. Basically, every time I hit `docker-build` it has to download the dependencies. Currently, I have two ways to fix this:
1. Push packages to the repo. But if I have ~30 microservices and most of them are 1 file program - it's a very poor experience.
2. `replace` and `require` create too many relative dependencies and it's very hard to maintain.

I expect 4 things from the `golang workspaces`:
1. `go mod` to know about other modules and don't try to download the local package from the internet. Maybe traverse up in the folder hierarchy and check if `go.work` exists?
2. you can create a clear dependency graph so it can be used for caching in Docker containers
3. functions like `go work tidy` and `go work download`, because now I have to enter every package and type `go mod` commands. I just want to tidy and download everything from one place, and because we already have a path to the modules in `go.work` file - it should be trivial.
4. functionality to reduce the number of similar dependencies (download one specific version and force to use it in every module in workspace). This was done by for example `turborepo`, `lerna`, and other monorepo management tools.

So now, the only thing why I'm using `go work` is to have a better codding experience in VsCode. Most of the negative comments that I hear from colleagues about golang are about poor modules/workspaces management. I think, `go work` is a very important feature to have and it's a very important direction to go.

**Comment by HeCorr** [comment]
I've faced this issue yesterday but my case is a bit different.. My `replace` statements are in the `go.work` file, which seemed to solve `go run` errors, but when attempting `go mod tidy` on any of the modules, it is indeed ignored.

The solution is simply to move the `replace` statements to the `go.mod` files as explained by Gbps, but keeping track of what package uses what internal package is a bit of a hassle. Is this considered a bug also?

**Comment by bcmills** [comment]
`go mod tidy` is intended to update the module's self-contained dependencies. It doesn't use the workspace because in general one may work on multiple independently-maintained modules in the same workspace, and if you're preparing an upstream commit you definitely don't want that commit to rely on unpublished local modifications.

`go work sync` is intended to update the modules within a workspace to reflect the dependencies selected in that workspace. That is probably what you want if you are working on a set of modules that are all maintained as a single unit.

Folks who are commenting here (and please bear in mind https://go.dev/wiki/NoPlusOne!) — have you considered `go work sync` for your use-case? If so, what does `go work sync` do (or not do) that makes it unsuitable for your use-case?

**Comment by AlmogBaku** [comment]
@bcmills I'm not sure what the intended workflow is, but I think the feedback here is that the DX is really not intuitive.

**Comment by burdiyan** [comment]
> Folks who are commenting here (and please bear in mind https://go.dev/wiki/NoPlusOne!) — have you considered go work sync for your use-case? If so, what does go work sync do (or not do) that makes it unsuitable for your use-case?

I wish there were `go work tidy` or something similar that would do the same as `go mod tidy` for each module in the workspace. Or maybe `go work sync` could have a flag for that, or even do it by default. This way one could easily manage a single repository with many modules, keeping everything in sync and tidy :) without having to switch to every single module directory to keep it tidy.

**Comment by lopezator** [comment]
There's an additional issue (I think) and I reproduced it cloning @bozaro repo. It you try to execute:

`go mod verify` it will complain with:

```
github.com/xmlking/go-workspace/cmd/app1 : missing ziphash: open hash: no such file or directory
github.com/xmlking/go-workspace/cmd/app2 : missing ziphash: open hash: no such file or directory
github.com/xmlking/go-workspace/cmd/app3 : missing ziphash: open hash: no such file or directory
github.com/xmlking/go-workspace/lib : missing ziphash: open hash: no such file or directory
```

There's something we can do to fix this?

**Comment by mlaventure** [comment]
`go work sync` does not work for me in that scenario either unfortunately. I have a `go.work` for a monorepo containing yet to be  published packages (the domain also does not exist yet). e.g.:
```
use (
    ./foo
    ./bar
    ./fubar
)

replace not-yet-created.com/foo/foo v0.0.0-00010101000000-000000000000 => ./foo
```

However that replace directive seems to be ignored by `go work sync`
```
$ go work sync
go: not-yet-created.com/foo/foo@v0.0.0-00010101000000-000000000000: unrecognized import path "not-yet-created.com/foo/foo": https fetch: Get "https://not-yet-created/foo/foo?go-get=1": dial tcp: lookup not-yet-created.com on 127.0.0.53:53: server misbehaving
```

**Comment by skelouse** [comment]
I've found that the problem with `go mod tidy` only happens when a sub-module does not exist in a package.

I've created a simple example to demonstrate the issue:  https://github.com/skelouse/go-work-test
- [Branch master,](https://github.com/skelouse/go-work-test) `go run .` and `go mod tidy` work as expected

- [Branch success](https://github.com/skelouse/go-work-test/tree/success), `go run .` and go mod tidy both work as expected.

- [Branch failing](https://github.com/skelouse/go-work-test/tree/failing), `go run .` works, but `go mod tidy` fails with:
```
github.com/skelouse/go-work-test/main imports
        github.com/skelouse/go-work-test/foo/newbar: module github.com/skelouse/go-work-test/foo@latest found (v0.0.0-20220809210529-8026891d438c), but does not contain package github.com/skelouse/go-work-test/foo/newbar
```

**Comment by skelouse** [comment]
@bcmills I'd be happy to contribute to this issue when a course of action is decided on.

**Comment by nathanblair** [comment]
[The documentation for workspaces](https://go.dev/doc/tutorial/workspaces)

> go.work can be used instead of adding [replace](https://go.dev/ref/mod#go-mod-file-replace) directives to work across multiple modules.

seems misleading here if not all of the `go` commands work with using the `go.work` as an override for `go.mod`.

I fully understand and for the most part agree that `go mod tidy` may not be the solution for addressing `tidy`-operations in a `go workspace` tree; a `go mod tidy` may modify a `go.sum` with unpublished references that when synced to a remote/upstream may include references that cannot be reached.

This could be a case of cleaning up/amending the documentation with further clarifications on what use-cases exactly a `go workspace` can solve.

**Comment by tliron** [comment]
I would like to very humbly suggest that we need a `go.work.sum` (or similar) file separate from `go.sum`, as we are dealing with two separate graphs of modules.

**Comment by nathanblair** [comment]
@tliron Coincidentally something (not sure what tool exactly - it was probably done under the hood from the VS Code Go extension) generated a `go.work.sum` file in some of my go workspaces.

**Comment by nine9ths** [comment]
I had unexpected behavior from `go get` related to go work. I'm concurrently developing a library in a directory and an application in a sibling directory. I did an initial commit on the library on main with just a go.mod file, pushed it to a remote (a private bitbucket instance), then started a develop branch. In the application I created a go.work that looked like:

```
go 1.19

use (
	.
	../library
)
```

I am also using vendoring in the application as part of a Docker build process. Everything was working pretty smoothly though, whenever I made a change to the library, I'd commit it,
then in the application run:

```
go get the-module-name/library@hash-of-latest-commit
go mod tidy
go mod vendor
```

then everything would be up to date. This would work regardless of whether I'd pushed the latest library commit to the remote.

At some point during development, I made a first pull request from the library's develop branch to main on the remote. Once I had done this, if I then tried to `go get` the library with a local only commit it errors with an invalid commit hash. I now must push the library commits to the remote in order to update them in the application's go.mod or vendor directory. This makes the go.work file irrelevant.

**Comment by bcmills** [comment]
@tliron, there is already a `go.work.sum` file that is populated if needed. (In many cases it is not needed — often the individual `go.sum` files in the workspace are sufficient to cover all of the relevant modules.) But that doesn't really seem relevant to this particular issue.

**Comment by bcmills** [comment]
@nine9ths, the behavior of `go get` for an unpublished commit is off-topic for this issue. (That's more closely related to #28835.)

**Comment by tliron** [comment]
@bcmills I did not know this and do not see this file in our environment. Could you point me to the documentation where `go.work.sum` is discussed? When is it populated? I assume that if a `go.work.sum` exists, then it would be used instead of `go.sum` for all operations?

Specifically to our topic, when I run `go mod tidy` does it update both `go.sum` and `go.work.sum` at the same time? Would there be a way to update only `go.work.sum`? That last ability would solve all our problems, which is why I raised the topic. During work, `go mod tidy` would fail and we know it would fail because ... we are in the midst of working on it. :)

We're having a lot of trouble with this `go.work` feature, like others here.

**Comment by bcmills** [comment]
@tliron, documentation for `go.work.sum` is #51941. Again, please keep discussion on the issue tracker specific to the issue.

**Comment by spekary** [comment]
@bcmills go work sync only works if you actually have a go.work file. Otherwise it errors.

I have a scripted install process that involves calling go mod tidy followed by a number of go install calls, all part of a continuous integration process. 

I was using go.work to be able to test the code in a working branch. However, go mod tidy attempts to grab the code from a release on a main branch. Normally, I want that, but when testing new code in a working branch, I don't. 

Trying to use go work sync would mean I would have to change the install scripts to do a go work sync if a go.work file was available, and go mod tidy if not. Difficult.

**Comment by betamos** [comment]
IIUC, `go mod tidy` downloads external modules declared in import statements and adds them to `go.mod`, but `go work sync` doesn't traverse source code? That's a fairly major tooling regression for workspace users. Are there other differences?

Independent of the original intent of `go mod tidy` and `go work sync`, it appears that the go toolchain disagrees with itself here. Either you are using workspaces or you're are not, or is there some inbetween state? When you type `go mod tidy` from muscle memory --  I'd assume the intent is "clean up in project the way it's organized" rather than "disregard my project structure and clean up as a standalone module".

As for motivation, many people *strongly prefer* monorepos for the overwhelming benefits they provide compared to the inconveniences and pitfalls of tightly coupled multi-repo projects. I'm sure Googlers can relate to this :) That said, I can't seem to find a definitive answer on whether monorepos is a first-class use case, and I've read the official docs, tutorials and design doc.

**Comment by Goldziher** [comment]
Its always amusing to discover how bad the go monorepo support is 😉. Frankly this is ridiculous - this thread is more than a year old, and instead of simply applying a fix the go maintainers attitude is to ignore it. Which is pretty much buisness as usual regarding anything DX related (see code coverage related issues etc.). 

> IIUC, `go mod tidy` downloads external modules declared in import statements and adds them to `go.mod`, but `go work sync` doesn't traverse source code? That's a fairly major tooling regression for workspace users. Are there other differences?
> 
> Independent of the original intent of `go mod tidy` and `go work sync`, it appears that the go toolchain disagrees with itself here. Either you are using workspaces or you're are not, or is there some inbetween state? When you type `go mod tidy` from muscle memory -- I'd assume the intent is "clean up in project the way it's organized" rather than "disregard my project structure and clean up as a standalone module".
> 
> As for motivation, many people _strongly prefer_ monorepos for the overwhelming benefits they provide compared to the inconveniences and pitfalls of tightly coupled multi-repo projects. I'm sure Googlers can relate to this :) That said, I can't seem to find a definitive answer on whether monorepos is a first-class use case, and I've read the official docs, tutorials and design doc.

A very commonsense thing to write. As are so many other posts in this thread.

**Comment by bcmills** [comment]
@Goldziher, simply quoting a previous comment and complaining about the lack of activity does not help to advance this conversation (see https://go.dev/wiki/NoPlusOne).

As noted in https://github.com/golang/go/issues/50750#issuecomment-1194322256, we haven't taken action on this issue because we don't have a clear net-positive change to make. If you want to move this conversation forward, please provide _constructive_ feedback on your concrete use-case. Thanks.

**Comment by BorzdeG** [comment]
@bcmills here is an example: https://github.com/BorzdeG/samples.go-and-flutter-grpc

If we execute `go mod tidy` in the `backend` directory, we will get an error importing the library that is in `public-api/go` and connected via `go.work`.

**Comment by wangmir** [comment]
@bcmills Maybe the difficulty to resolve this is related to the nature of go.mod, and go.work things.

How about just add link to the go.work from go.mod to override the go.work's `use` or `replace` directives first before handling go.mod. 

So, we can ensure that, in case of we cannot find go.work from the link (relative path maybe), then just follow the version and URL, but when we have go.work, then we can follow the path of go.work.

**Comment by BorzdeG** [comment]
Maybe add an `go work tidy` command similar to `go mod tidy` , but understanding `go.work`?

**Comment by wangmir** [comment]
@BorzdeG I think `go work tidy` is also not a complete solution because in that case, the module should be tightly coupled with go.work. If we can just point out go.work from go.mod, the go module can be still managed independently, but just reference the go.work to get proper library only.

**Comment by wangmir** [comment]
@bcmills Hi, are there any discussion channels for this issue?

**Comment by Nikola-Milovic** [comment]
From this thread, it seems that this is the intended functionality. This makes workspaces so much less useful than they could be, any updates on discussions, is this being mentioned in some proposals?

**Comment by hellt** [comment]
I wonder if I miss something here, but I intended to use workspaces as a more cleaner "replace" alternative when working on a module that imports a package from my own unpublished fork.

E.g. I am working on a `github.com/user/project1` and it imports `github.com/user/project2`.

The `project2` is not published yet and it is in my go.work file imported by a local path.

But then if I need to import new public dependencies in `project1` I need to run `go mod tidy` to refresh the go.sum and process relative imports, and this fails, since `go mod tidy` tries to find `project2`.

**Comment by jason-apolline** [comment]
> I wonder if I am missing something here, but I intended to use workspaces as a more cleaner "replace" alternative when working on a module that imports a package from my own unpublished fork.
> 
> E.g. I am working on a `github.com/user/project1` and it imports `github.com/user/project2`.
> 
> The `project2` is not published yet and it is in my go.work file imported by a local path.
> 
> But then if I need to import new public dependencies in `project1` I need to run `go mod tidy` to refresh the go.sum and process relative imports, and this fails, since `go mod tidy` tries to find `project2`.

This is exactly how I was expecting it to work, and I have the same issue. I've ended up just using standard go.mod files with replace directives and stripping them out with sed when building my docker images. Workspaces sounded like it would be ideal for this scenario, but sadly not.

**Comment by dkrieger** [comment]
Since the go team doesn't seem interested in addressing this issue, perhaps a tool that configures a local proxy using the go.work file could work?

It is very strange that the default assumption in a module that is nested in a workspace wants to depend on pinned published versions of other modules in the same workspace, especially when the language itself was developed in a monorepo to avoid this state of affairs. Gazelle also avoids this state of affairs, but incompletely because it uses go.mod files which go mod tidy cannot correctly manage. I can only assume there is a private tool or fork of the go command used in place of the public `go mod tidy` command that has removed any motivation to upstream a fix.

The messaging from users has consistently been that this behavior doesn't actually serve any use cases that could not still be served by simple checks for workspace context. Even if you want to publish derivative single module repositories in public mirror repos, nothing about changing go mod behavior within the context of a go workspace would break that use case.

**Comment by dezren39** [comment]
> Since the go team doesn't seem interested in addressing this issue, perhaps a tool that configures a local proxy using the go.work file could work?
> 
> It is very strange that the default assumption in a module that is nested in a workspace wants to depend on pinned published versions of other modules in the same workspace, especially when the language itself was developed in a monorepo to avoid this state of affairs. Gazelle also avoids this state of affairs, but incompletely because it uses go.mod files which go mod tidy cannot correctly manage. I can only assume there is a private tool or fork of the go command used in place of the public `go mod tidy` command that has removed any motivation to upstream a fix.
> 
> The messaging from users has consistently been that this behavior doesn't actually serve any use cases that could not still be served by simple checks for workspace context. Even if you want to publish derivative single module repositories in public mirror repos, nothing about changing go mod behavior within the context of a go workspace would break that use case.

https://github.com/goproxy/goproxy

**Comment by mdepot** [comment]
As you devs work out how to fix this, please keep this use case in mind.  One of the reasons to use a workspace is the scenario where we want to build multiple binaries, and those binaries share some common code or other resources.  A client-server pair of binaries using a common protobuf makes a good example.  Here one might use a layout like this:

```
myapp
   go.work  
   bin
      client.exe
      server.exe
   cmd
      client
         client.go
      server
         server.go
    proto
       myapp.proto
    pkg
       common1
          common1.go
       commmon2
          common2.go

```
The desire to use a workspace here is in order to work on these two binaries as a single unit.  This is because:
 * They are using a common protobuf, so client and server must stay in sync.
 * They are sharing some common packages.
 * They may be version controlled in a single git repo, because of the above.
 * We may want to have CI build and release the client and server together.
 * IDE convenience: We should not need to have two separate copies of VSCode open when working on a client-server app, therefore we are using a workspace containing both binaries.

When working under this use case, the naive go newcomer may expect that running go mod tidy at the top of the workspace (as well as some other commands) should produce the same result as if they ran it separately inside lower directories.  This is because they are thinking in the mindset that these are two separate apps.  Granted that's not how the existing tool options are supposed to work today, but in the long run it would be nice to have options in the tools that support using a workspace for this use case, to maintain what are essentially independent binaries that just happen to need to be maintained together as a single unit.

Whatever the answer ends up being, it would be good to explain the best way to manage this type of scenario somewhere in the workspace docs as well.

**Comment by mdepot** [comment]
Depending on what a user's expectations are, this may have some overlap with https://github.com/golang/go/issues/40302

**Comment by matloob** [comment]
@mdepot What's the use case for the client and server being in different modules from each other?

**Comment by dkrieger** [comment]
> @mdepot What's the use case for the client and server being in different modules from each other?

can we not XY-problem this and take as given that it is a valid use case to have a client and a server be in different modules, especially considering it is commonplace for them to be in entirely different repos?

**Comment by BourgeoisBear** [comment]
@matloob: i'm in the same situation as @mdepot.  In many cases, the client is free / open-source while the server is proprietary / closed-source.  There are also cases where one side has far more dependencies than the other (i.e. why force people to pull a fat Electron client when they only want the server side that's strictly files & sockets.

**Comment by Perusae** [comment]
> Whatever the answer ends up being, it would be good to explain the best way to manage this type of scenario somewhere in the workspace docs as well.

We also use a similar structure for the above mentioned reason.  When running Go tidy from the root, it took us by surprise that it ignored the go.work. 

A quick fix was running:

    for f in $(find . -name go.mod)
          do (cd $(dirname $f); go mod tidy)
     done

But that is just a temporary solution, for a missing feature.

**Comment by bcmills** [comment]
@dkrieger, generally we expect “dependencies on other packages' source” to mean something that can be fetched as a module, from a (public or private) GOPROXY or version-control host. It isn't obvious why one would have dependencies that can't at least be redirected for `go get` using an `insteadOf` directive in the user's `.gitconfig`.

**Comment by bcmills** [comment]
@BourgeoisBear
> There are also cases where one side has far more dependencies than the other (i.e. why force people to pull a fat Electron client when they only want a server side that is strictly files & sockets?).

That was the motivation for adding [module graph pruning](https://go.dev/ref/mod#graph-pruning) in Go 1.17. Are there cases of this sort where graph pruning is ineffective? If so, can you give more detail about the concrete problems (ideally as a separate issue)?

**Comment by dkrieger** [comment]
> @dkrieger, generally we expect “dependencies on other packages' source” to mean something that can be fetched as a module, from a (public or private) GOPROXY or version-control host. It isn't obvious why one would have dependencies that can't at least be redirected for `go get` using an `insteadOf` directive in the user's `.gitconfig`.

The obvious case is when you have a monorepo with multiple go modules, orchestrated with that other Google tool bazel/blaze.

**Comment by bcmills** [comment]
@dkrieger, can you give some more detail about why the monorepo is structured that way? (We're trying to better understand the underlying use-cases so that we can address them holistically.)

We're familiar with https://github.com/googleapis/google-cloud-go, which IIUC uses multiple modules so that they can tag its various APIs at different levels of stability (for example, have some APIs stabilized at v1 while others are still at v0). But the discussion on this issue has been mostly focused on dependencies that don't have any versions published at all, which would imply that that's not the same underlying motivation.

**Comment by BourgeoisBear** [comment]
@bcmills, "pull" as in git/hg of a fat repo with large assets (like a GUI with many rasters/videos) when I don't need it.  No complaints on compile time or binary sizes.

**Comment by bcmills** [comment]
Hmm, here's a thought. For #50603, we're going to need the ability to inspect the local repo containing a given repository to determine its version information.

In theory, `go work sync` could do the same thing for every main module in the workspace, so that if some module in the workspace depends on it that dependency could be updated to refer to the version actually found in the workspace.

But there is a bit of a chicken-and-egg problem: the version of a checked-out repo in the workspace depends on its commit hash, which depends on the contents of its files. If there is a cycle in the requirement graph (which is normally allowed), then `go work sync` would become unstable: each time it writes out a `go.mod` file in one module, the versions it records in the other `go.mod` files would be updated to the new hash, causing an additional `go.mod` write. The only ways to reach a fixed point are:
1. Decouple the actual versions written to the `go.mod` files from the versions found in the workspace, such as by resolving upstream commits.
2. Have `go work sync` intentionally _remove_ references to the workspace modules from the `go.mod` files, presumably to be filled in later.
3. Have the user explicitly specify versions to assign to the workspace modules, instead of trying to infer them from the repo.

**Comment by dkrieger** [comment]
@bcmills let me start with a language agnostic description of why we would structure a monorepo in this way, then I'm happy to field any go-specific follow-up questions you might have.

At our company (and I'd suggest most companies), most of our code is not public. Our deployable artifacts are mostly containers that we deploy to kubernetes.

For some given deployable slice of the codebase, we may deploy it as one or more services, and a given chunk of code may appear in one or more containers -- in a common scenario where it appears in multiple containers, we may have a fatter container that aggregates multiple domains (let's say each domain is represented by one grpc service, and we pack those into single networked service, i.e. many:1 domain:service) and a smaller container that is a microservice (i.e. 1:1 domain:service). This enables sophisticated traffic shaping when both are deployed, but in more simpler scenarios it means we can periodically move predictable-traffic and/or high-reliability domains into the aggregate service, and move bursty and/or low-reliability domains into their own microservices -- basically, we preserve domain boundaries and the flexibility of deploying as a microservice or in an aggregate/monolith as we see fit, balancing performance, complexity, efficiency, and reliability over time.

At lower levels, we may have various utility libraries that get used by many deployable slices of the codebase.

Now, for all that private code, we don't really care about publishing these for consumption _from other repos_, whether inside or outside the organization. As such, we don't have to worry about versioning in the traditional sense -- any commit on the trunk branch (or, by convention, any merge commit) can be built and deployed, and the commit hash is the "version" for everything. Because everything we do is in the monorepo, we not only know every consumer for X service, but they share the same build and IDE context. We can make breaking changes to a service and update every consumer in a single commit if we want to (in practice we'd make a backwards compatible change followed by cleaning up the deprecated API when talking about services, but in the case of libraries we can do it all in one shot).

Still, we want strong boundaries between different parts of our codebase, and the universally strongest boundary is to structure it as the unit a given language uses for external dependencies -- in node, this is a package; in go, this is a module. We don't actually want to put package registries in the middle of this though.

When I look at my string util package ("module" in go), I want to be able to look at the go.mod file and see just the dependencies for that, not polluted with a single line that has to do with some consumer of the string util go module, or some other completely unrelated go module. If it doesn't depend on any other go modules in the monorepo, it could be built with just the standard go toolchain.

For higher level go modules in the monorepo (e.g. a networked service), if I don't want them published for consumption outside the monorepo, they will have dependencies on external libs and some other internal modules in the monorepo. I'd like to be able to `go mod tidy --workspace` and not have it barf, behaving how it currently does for external libs, but for other modules in the workspace (as defined in the go.work file), it will point to those local sources. This allows for making logical changes to module apis in a single commit, updating consumers that would otherwise break, without going through the toil of publishing each update in the dependency graph first. With bazel, every part of thus workflow works except for the `go mod tidy` portion, because `go mod tidy` doesn't understand how to handle this.

**Comment by dkrieger** [comment]
@matloob Bazel with rules_go and gazelle lets you configure gazelle to source your workspace-level go dependencies from the union of every go.mod in your bazel workspace. A bazel workspace maps to a go.work, if you are using go modules and workspace. The only pain point is the missing ability to have go mod tidy create a go.mod for you that uses your unpublished/unversioned sources for other modules in the same go/bazel workspace

**Comment by matloob** [comment]
@dkrieger I didn't realize that Bazel supported that.

I want to try to step back a bit. The reason I'm so hesitant about supporting this is that we don't want the tooling to encourage folks to use multiple modules when they can work with a single one. The motivation that I understand for having a multi module monorepo is that the modules are distributed separately from each other. That is: the modules appear in a monorepo, but they are also dependencies of other external modules. For those cases go mod tidy should work fine because each module is able to stand on its own as a dependency of an external project.

So I want to understand your use case better. I see that you mentioned the string util libraries. Are those depended on by other modules outside of your monorepo? Are they built separately from the rest of your monorepo?

**Comment by dkrieger** [comment]
@matloob 

> The reason I'm so hesitant about supporting this is that we don't want the tooling to encourage folks to use multiple modules when they can work with a single one.

I have a couple responses to this:

1. an optional cli flag does not encourage its usage, and the docs can express (non-)recommended use cases. In a pnpm monorepo, only when you add the `--workspace` flag when adding a dependency will it prefer the local workspace sources (vs some pinned version available on the registry) when adding an intra-workspace dep.
2. I think borrowing the python mantra of "we're all adults" is appropriate here. Ergonomics and defaults are the appropriate ways to signal recommended patterns, not capabilities. In python, prefixing with an underscore signals "this isn't meant for external use", but you can still do it if you feel you have a good reason to; this works out really well in practice

> The motivation that I understand for having a multi module monorepo is that the modules are distributed separately from each other.

For all intents and purposes, I'd say that the use case I'm describing _does_ involve distributing modules separately from one another; whether they're distributed in the conventional go way or not, whether they're published to a public registry (including public github repo) or not -- these are downstream decisions. In all cases, they're possibly (and in practice, often) _developed_ in lock step, and contracts can be changed in backwards incompatible ways in a single atomic commit. Again, the main effect is eliminating the toil of traversing your dependency graph and the logical change you're wanting to make becoming distributed over the dimension of git-commit-time

The result is much cleaner than having a huge module with packages that have no logical relationship with one another. Every module I create tends to consist of several packages, and those packages all pertain to the logic of that module. The 2nd level of ordering that modules provide is incredibly useful, as it (1) makes it easier to understand the transitive closure of dependencies of any piece of code without noise and (2) makes it easier to understand the interrelated packages that exist in that module. This is just as true whether or not I publish my modules in a way that they can be retrieved via `go get` in some external repo/organization.

**Comment by nemith** [comment]
> That was the motivation for adding [module graph pruning](https://go.dev/ref/mod#graph-pruning) in Go 1.17. Are there cases of this sort where graph pruning is ineffective? If so, can you give more detail about the concrete problems (ideally as a separate issue)?

When downloading a module the entire repo is checked out.  So if you have a single go.mod for a monorepo you easily run into an issue where you can hit the limits of size of repo (I can't remember what the limits are but it's hardcoded into the go command).   Using submodules does inform git to ignore directories. 

I am specifically running into this problem with trying to add new modules to an existing repo.  New module is created, needs to be referenced by other modules in the monorepo but `go get` and `go mod tidy` don't work as the module doesn't exist upstream yet.  This is breaking a normal workflow for coding and means we have to commit a dummy `mod` first (and get code review on it) before hacking further.

Replace works here, but sorta defeats the purpose of `go.work`.

**Comment by spekary** [comment]
> means we have to commit a dummy mod first (and get code review on it) before hacking further
Yep, that is how I had to work around it. I don't think it has anything to do with limits on the size of a repo though.

**Comment by nemith** [comment]
> > means we have to commit a dummy mod first (and get code review on it) before hacking further
> 
> Yep, that is how I had to work around it. I don't think it has anything to do with limits on the size of a repo though.

What i mean is that I am using (or rather want to us) multiple modules to get around the size of my monorepo vs just having a single `go.mod` for the entire repo.

There are two problems with large monorepos and a single `go.mod` when it comes to size.  One is just download size, it takes a long time for consumers of one project in the the monorepo to download and cache the entire thing.  The other is hard limits.

For all download type it looks like it defaults to 10<<20  (or around 524MB) which is a lot for a repo, but we use git-lfs which is resolved here and includes any binaries you may want to stuff into your monorepo:

https://github.com/golang/go/blob/master/src/cmd/go/internal/modfetch/codehost/codehost.go#L35 
https://github.com/golang/go/blob/master/src/cmd/go/internal/modfetch/proxy.go#L435
https://github.com/golang/go/blob/master/src/cmd/go/internal/modfetch/codehost/vcs.go#L465
https://github.com/golang/go/blob/master/src/cmd/go/internal/modfetch/coderepo.go#L1087


Submodules is a great way to solve this, but because of resolving versions and the chicken-and-egg problem they are hard to work with and hard for my coworkers to stomach.

**Comment by dkrieger** [comment]
An important underlying theme here is that monorepos are not monolithic codebases, nor are they distributed eventually consistent codebases where you necessarily version and publish each part to some registry. They are no less modular than multi-repo/polyrepo -- to the contrary, they promote mindful modularity by reducing the incidental disincentives to modular design. That they don't encourage/require long-lived support branches in the form of versioning in many circumstances is frankly unrelated to modularity.

If we can agree that this is at least a defensible position, let's try to shift the focus to how we can decouple go module tooling behavior from opinions for/against monorepo adoption without breaking userland.

**Comment by rsc** [comment]
"go mod" commands are scoped to modules; there is no workspace in go mod sub-commands. This is why "go work vendor" is different from "go mod vendor". So "go mod tidy" should not know about go.work any more than the other commands.

It seems like "go work sync" is what we should be focusing on. What does it do or not do that is inappropriate?

**Comment by dkrieger** [comment]
@brianbraunstein 

> Possible Solution 2: Clarify the purpose of workspaces

Respectfully, this is not a solution, it's a temporary mitigation at best. Not supporting atomic updates spanning multiple workspace modules defies the purpose of a workspace in the general sense of the term, as it is observed across languages and monorepo build tools. Whether we get `go mod tidy --workspace`, `go work modtidy`, or some other cli for this use case has no bearing on whether it is a sensible use case to support or not. If we can agree it satisfies a standard of "plausibly sensible", that should be enough to move forward, at which point the finer details of the appropriate cli API can be worked out.

**Comment by atdiar** [comment]
I don't know whether it is related but one issue that I have encountered is, while building something similar to gonew, when downloading a module in a workspace, the go.work file needed to be updated manually.

Perhaps that go work sync could also work bottom up (via a flag?) to add this module/directory to the module list in the go.work file?

If that makes sense.

**Comment by dpifke** [comment]
If a different command should be substituted for `go mod tidy` when using workspaces, I think we should update the various places where running other commands instruct the user to run it:

https://cs.opensource.google/go/go/+/refs/tags/go1.22.0:src/cmd/go/internal/modget/get.go;l=1781
https://cs.opensource.google/go/go/+/refs/tags/go1.22.0:src/cmd/go/internal/modload/help.go;l=53
https://cs.opensource.google/go/go/+/refs/tags/go1.22.0:src/cmd/go/internal/modload/init.go;l=648
https://cs.opensource.google/go/go/+/refs/tags/go1.22.0:src/cmd/go/internal/modload/init.go;l=651
https://cs.opensource.google/go/go/+/refs/tags/go1.22.0:src/cmd/go/internal/modload/load.go;l=2023
https://cs.opensource.google/go/go/+/refs/tags/go1.22.0:src/cmd/go/internal/modload/load.go;l=2029
https://cs.opensource.google/go/go/+/refs/tags/go1.22.0:src/cmd/go/internal/modload/load.go;l=2032

(This might also be a good way to audit `go mod tidy` use cases and make sure they are covered by whatever is the equivalent workspace command.)

Being told by the tooling to run `go mod tidy`, and having that fail, is what brought me to this issue originally.  I've since given up on using workspaces and instead use `replace` directives in `go.mod`, along with a hacky Git post-commit hook to filter those out when committing.  (I'm probably holding it wrong.)

**Comment by ohir** [comment]
@matloob 

> So I want to understand your use case better.

Many use cases were described by participants here.  Mine is that I do not want to publish dummy stub modules before I can start to work. I can buy __rsc__ explanation why ```go mod tidy``` should not know about ```go.work```.  But certainly ```go work``` should be able to recursively tidy all ```use``` pieces declared in ```go.work``` workspace, akin to ```go work sync```.   Thus please consider:

``` go work tidy```

that does not try to reach to the unpublished modules – ie. ones ```use```d in ```go.work``` with a ./ path.

**Comment by matloob** [comment]
@dpifke Thanks, that's a useful perspective. We don't want our tools confusing users with the instructions they give. I'm interested if solving the error messages is the primary reason most people want 'tidy' for these use cases.

@brianbraunstein I don't think 'tidyness' is the right concept here. I don't think we should call a `go.mod` file tidy if it can't be used (alone) to build a module. If `go work sync` doesn't work we'll need to think of some other clean concept we can use to express that though the modules don't work standalone they can be used in a workspace concept.

@ohir I don't know if `tidy` is the right concept here. Why wouldn't `go work sync` work for your use case.

@dkrieger I don't think we should frame this discussion around what the concept of a "Workspace" is in other language ecosystems. We want to build what make the most sense for the Go module system.

**Comment by ohir** [comment]
@matloob

> Why wouldn't go work sync work for your use case

It still reaches to the net even if it has everything replaced or internal.  It should not.

Repro tree attached (usage: `patch -p0 < goworksync_repro.patch`) 

[goworksync_repro.patch](https://github.com/golang/go/files/14693896/goworksync_repro.patch)

**Comment by matloob** [comment]
@ohir 

When I disable the network your repro case doesn't produce an error when the two replaces in the go.work are there. But if I remove them I do get an error. Is that what you're seeing too? If not, what version of Go are you running?

**Comment by ohir** [comment]
@matloob 

There is no problem with `sync` if workstation network (local) is down.  Problem arises when the workstation network is on but module repo is not exposed – either due to the firewall rules, or because it hasn't been really set up. 
 
Since `sync` tries to contact `fairbe.org` domain looking for the module repo it hangs for a longer while (till Dial timeouts).
With real tree having dozens of modules it takes more noticeable time.

This is a wider problem: `go get` also goes astray on a repo with unpublished/unreachable module:
It errs with `[...] Get "https://fairbe.org/moda?go-get=1": net/http: TLS handshake timeout`.

The only working solution for now is to replenish all `go.mod` files with respective replaces, too.  What nullifies all, or almost all, benefits of the _workspace mode_.

------
Summary:
Workspace _can not_ be reliably set up for a project that is even partially isolated, ie. it contains an unpublished/unreachable module(s).  For the `go work sync` and `go get` to work seamlessly everything in the project, _including isolated parts_, MUST be reachable over the network - or be `replaced` directly in impoter's `go.mod` file.

**Comment by ohir** [comment]
@matloob 
FYI:  neither you can easily vendor an unpublished module in the workspace.
 
After adding `vendor/btea` module to the workspace (unpublished `fairbe.org/btea`) then trying to import/use it resulted in a few minutes of delay in the IDE (awaiting network timeouts) then `gopls` throwed at me a full bucket of errors.  Interestingly — import errors in mostly unrelated modules.  Not even stdlib pieces could be found.

Filtered excerpts follows:

[cmd/smth imports `fairbe.org/btea` module for `btea.FromBELE`]
`: "undefined: FromBELE"` <= this should come from vendor/btea, it did not.  Until replaced explicit in the `cmd/go.mod`

[other module - imported by cmd/smth but NOT importing `fairbe.org/btea` at all
```
: "could not import golang.org/x/crypto/chacha20 (missing metadata for import of \"golang.org/x/crypto/chacha20\")",
: "fairbe.org/btea@v0.0.1: unrecognized import path \"fairbe.org/btea\": https fetch: Get \"https://fairbe.org/btea?go-get=1\": net/http: TLS handshake timeout",
: "could not import fmt (missing metadata for import of \"fmt\")",
: "could not import crypto/ecdh (missing metadata for import of \"crypto/ecdh\")",
: "could not import crypto (missing metadata for import of \"crypto\")",
: "could not import golang.org/x/crypto/chacha20 (no required module provides package \"golang.org/x/crypto/chacha20\")",
```

To sort this mess out an explicit replace had to be added to the (cmd's) go.mod `replace fairbe.org/btea v0.0.1 => ../../../vendor/btea`.  

I hear "its a gopls issue".  No, it is not just `gopls` – it is a problem of workspaces still being a patchwork on the side of other tools mostly being unaware of the go.work replace directives (@rsc).
 
I would like my "Summary" warning to be included in the docs of the workspaces for the time being - to save headaches for others trying to add some isolated module to their workspace.  Thanks.

**Comment by GustavoStingelin** [comment]
Hey everyone,

If you declare the local shared module only with the folder name, you can import it directly in the other module Go files without needing a remote go.mod.

By example:
```
myproject/
├── go.work
├── api/
│   ├── cmd/
│   │   └── main.go
│   └── go.mod
└── shared/
    ├── go.mod
    └── middleware/
        └── middleware.go


$ cat go.work
    go 1.22.1
    
    use (
            ./api
            ./shared
    )

$ cat shared/go.mod 
    module shared
    
    go 1.22.1
    
    require (
       ...3rdparty/lib
    )

$ cat api/cmd/main.go 
    package main
    
    import (
            "shared/middleware"
    )

$ cat api/go.mod 
    module api
    
    go 1.22.1
    
    require (
       ...3rdparty/lib
    )
```

**Comment by dkrieger** [comment]
> Hey everyone,
> 
> If you declare the local shared module only with the folder name, you can import it directly in the other module Go files without needing a remote go.mod.
> 
> By example:
> 
> ```
> myproject/
> ├── go.work
> ├── api/
> │   ├── cmd/
> │   │   └── main.go
> │   └── go.mod
> └── shared/
>     ├── go.mod
>     └── middleware/
>         └── middleware.go
> 
> 
> $ cat go.work
>     go 1.22.1
>     
>     use (
>             ./api
>             ./shared
>     )
> 
> $ cat shared/go.mod 
>     module shared
>     
>     go 1.22.1
>     
>     require (
>        ...3rdparty/lib
>     )
> 
> $ cat api/cmd/main.go 
>     package main
>     
>     import (
>             "shared/middleware"
>     )
> 
> $ cat api/go.mod 
>     module api
>     
>     go 1.22.1
>     
>     require (
>        ...3rdparty/lib
>     )
> ```

Now try running `go mod tidy` in myproject/api

**Comment by GustavoStingelin** [comment]
@dkrieger all looks good here! No errors or warnings.

![image](https://github.com/golang/go/assets/59903150/8a2742b5-ce94-452f-85da-c3421cf6bf3f)

**Comment by GustavoStingelin** [comment]
> > Hey everyone,
> 
> > 
> 
> > If you declare the local shared module only with the folder name, you can import it directly in the other module Go files without needing a remote go.mod.
> 
> > 
> 
> > By example:
> 
> > 
> 
> > ```
> 
> > myproject/
> 
> > ├── go.work
> 
> > ├── api/
> 
> > │   ├── cmd/
> 
> > │   │   └── main.go
> 
> > │   └── go.mod
> 
> > └── shared/
> 
> >     ├── go.mod
> 
> >     └── middleware/
> 
> >         └── middleware.go
> 
> > 
> 
> > 
> 
> > $ cat go.work
> 
> >     go 1.22.1
> 
> >     
> 
> >     use (
> 
> >             ./api
> 
> >             ./shared
> 
> >     )
> 
> > 
> 
> > $ cat shared/go.mod 
> 
> >     module shared
> 
> >     
> 
> >     go 1.22.1
> 
> >     
> 
> >     require (
> 
> >        ...3rdparty/lib
> 
> >     )
> 
> > 
> 
> > $ cat api/cmd/main.go 
> 
> >     package main
> 
> >     
> 
> >     import (
> 
> >             "shared/middleware"
> 
> >     )
> 
> > 
> 
> > $ cat api/go.mod 
> 
> >     module api
> 
> >     
> 
> >     go 1.22.1
> 
> >     
> 
> >     require (
> 
> >        ...3rdparty/lib
> 
> >     )
> 
> > ```
> 
> 
> 
> it only makes go mod thinks the module is in the std. compiling will fail.

I have this project in production.

**Comment by rufreakde** [comment]
Since this thread seems to summarize many different opinions and ideas as to what to change for the `go.work` behaviour let me add my two things I would love to have for "early development" usability.

# 1 dependency management for "initial setup"
I personally hate it that if I work on several new submodules i would need to release an initial version first for all of them to be able to reference them correctly on a main go.mod file. Otherwise the IDE always complains...

why not allow to reference inside go.mod something like a keyword to emphasize that this is handled by go.work file and is not maintained currently (yet) against a stable version.

something linke:
```go
require https://github.com/google/uuid vWorkspace
```

linters can also work with such a keyword showing the warning for this... builds could fail initially when keyword is found.
And for initial setup you can finish a submodule with main module tests before pushing or releasing it.

Currently my workaround is:
 to push it to a branch and use the branch as a version this could help others as well currently maybe.
 
 With this keyword also go mod tidy would just work. Because the expectation of "this is not a final version" is fulfilled and IDE compiler etc still works without extra effort of manual branch versioning.

# 2 extra work for submodule tidyness
The second point I read here somewhere in the chat already and it is the need to "cd" into all directories to "go mod tidy" all submodules.

Workaround is to edit the make file:
```sh

tidy: ## Tidy go modules
	go work sync #sync root
	cd apis/A/v1alpha1 && go mod tidy # tidy all submodules
	cd apis/B/v1alpha1 && go mod tidy
        ...
        go mod tidy # tidy main module
```
So here I would love an option to do something like:
```sh
go mod tidy --recursive #(and if there is no main module it should still look into subfolders...)
```
I think the "featureset" that is available is usable but with manual effort which is confusing as most of the things where also possible with manual effort without the `go.work` file and this spikes confusion.

### conclusion:
IMHO this two small changes would come a long way in terms of "Developer happiness"

**Comment by leemeichin** [comment]
To add my feedback, we've been using go in our monorepo and I think the confusion fundamentally boils down the decision that workspaces are an override for local development, but according to guidelines shouldn't be production, and therefore `go mod` and `go work` are orthogonal.

We have a setup like:

```
./services/...
./packages/go/...
```

We used to have a `go.mod` at the top level, but that made every package share the same set of dependencies. Creating a workspace seemed like a nice approach (like yarn workspaces in JS for monorepos), and then we'd set up each package in `services` and `packages/go` with their own `go.mod` file. The result is that we have to micromanage it with `replace` directives now, and those can only be added by hand or through the CLI.

We also tried deleting the workspace and running `go mod tidy` again, and it doesn't understand imports from other parts of the same repo, yet the code builds successfully.

From my perspective, I feel like the tooling is unintuitive and is probably the most 'magical' part of the language, considering how methodical and obvious the rest of the language is.

**Comment by li-jin-gou** [comment]
Any progress?

**Comment by li-jin-gou** [comment]
I've encountered a similar problem. the current solution is write a shell to `go mod replace` first and then `go mod tidy`  and removing  `go mod replace`.  Hope GO Support `go mod tidy` Ignore go.work.

**Comment by EduartePaiva** [comment]
As someone who is starting with Go, and learning micro-services, I want to share my perspective, on this workspace thread.

Here's my user case  that I was working on with workspace:

- created module micro-service-1.
- created module micro-service-2.

- created now a utils module, currently micro-services-1 uses it.
- created a gRPC module, it have generated code that allows ms-1 and ms-2 to talk to each other.

Now I want to build docker images for ms-1 and ms-2.

what would be my advantage for using workspace in this scenario? 
- Faster download times for building images, for example, ms-2 don't use utils module so it don't need to copy it and download it's dependencies.
- Better mental model for external packages.
- from a newbe perspective this looked as the right way to do things.

now here's the problem that is emerging:

- having to manually tidy things
- now for tidy to work I have to do workaround with this: removing link style from module names, publishing as a package, using replace (learned about this now)

Too much trouble, probably I'll just stop using workspace and forget the possible advantages

## My suggestion for a "go mod tidy --workspace", or "go work tidy".

It should recursively go to workspace modules and tidy them too.

For example, if inside ms-1 I run the supposed "go work tidy", it should see that it's using the utils module and recursively tidy utils too. 

If this behavior is debatable put a flat for it e.g. "go work tidy --recursive".

**Comment by matloob** [comment]
@EduartePaiva I'm wondering why you didn't place all your packages in the same module? That is the happy path. And then no workspace is required.

To others on this issue: I would be like to speak to people about the requirements they have that push them towards multi-module repos with go.work files. I think if we could better understand those requirements we would be better equipped to see how to satisfy them.

Please let me know if you're interested in talking about your use case.

**Comment by EduartePaiva** [comment]
@matloob being honest I was following a microservice course and the teacher started using workspace to separate each microservice. At the time I didn't questioned why, just thought that it was the way to do things.

**Comment by jurienhamaker** [comment]
@matloob In my case I am rewriting several discord bots to go. However they have some shared code which I am putting in their own package & the services themselves.

[You can check it out here](https://github.com/jurienhamaker/Yugen/tree/go)

Now I got a dockerfile for each and I copy the mod file for the app, then the shared mod file and download them seperately etc.

**Comment by KevinFairise2** [comment]
Hi,
Working on https://github.com/DataDog/datadog-agent, we decided to split our repository in several modules to be able to import part of the agent code without importing the whole code base. It definitely worked but brought some challenge on the tooling to manage the multi module repository.
We recently introduced `go.work` and added it to the repository. It improves things, because we are now able to keep dependencies version synchronised between the modules and it simplifies running test in the different modules. But we have not been able yet to remove all the `replace` from the `go.mod` files because of we the things mentioned in this issue.

**Comment by GustavoStingelin** [comment]
> ## My suggestion for a "go mod tidy --workspace", or "go work tidy".
> 
> It should recursively go to workspace modules and tidy them too.
> 
> For example, if inside ms-1 I run the supposed "go work tidy", it should see that it's using the utils module and recursively tidy utils too. 
> 
> If this behavior is debatable put a flag for it e.g. "go work tidy --recursive".
> 

I did this using a Makefile to handle the tidy

**Comment by matloob** [comment]
@EduartePaiva Ah, got it. Yeah, in general multi-module repositories should only be reached only if a single repo repository isn't working for your use case.

@nemith Do you have modules in your workspace that aren't `go get`-able? And if so, what's the constraint that prevents them from being `go get`-able? My understanding is that that is when this issue occurs- when there are modules in your workspace that can't function independently.

@jurienhamaker Are the bots in separate modules to ease deployment similar to the problems @nemith mentioned?

@KevinFairise2 What were the downsides of importing the whole codebase? Was it to restrict the size of dependencies that were brought in with the imports? Also, are each of the modules independently go-gettable? If you're able to do a call I'd definitely be interested in talking about your use case.

**Comment by jurienhamaker** [comment]
@matloob The bots itself can function independently as long as they are compiled together with the shared module. The shared module is their only dependency. 

The dockerfiles in the repository show basically that, I don't add service y & z to the docker container of x.

**Comment by nemith** [comment]
@matloob I am going to go off on memory here (maybe I will try to reproduce it if i get time later).   But as I remember everything worked fine until I added a new module that wasn't yet committed to the repo.  

Despite being listed in go.work no other modules could try to use it as it would try to contact the repo (github in my case) to get information on the imported repo instead of just looking locally first.  

I guess that makes it not `go get`able but that was just because of order of operations and not desired end state. 

If i remember right there were workarounds but all of it added too much cognitive load to the team who's job was to write software and not understand the rube goldberg building machine as much as I could.

**Comment by majelbstoat** [comment]
@matloob one consideration is that Go is not necessarily the whole picture for any given project. We have a product with a front-end in Typescript, multiple backend services in Go, and an app in Swift. They all communicate using GRPC, and depend on a single common set of protobufs, from which we generate types, clients, and server stubs.

Keeping all the protos, services, apps, and frontends in one repo allows us to change protobufs, regenerate all the client and server stubs, and commit the consequences to the whole system all in one go. in particular, we can understand if anything is going to break before changes are merged to main. New versions of affected services and apps can also be automatically deployed based on changes to protos on which they depend.

All the go code being in one repo is a subset of that need, and similar to what others have said: common grpc (and other) code in `lib/go/src` and service code is `services/service-one`, `services/service-two` etc.

To make this work, we commit go.work to the repo (even though that's not its intended use). go.work and go.mod look like:

go.work:
```
use (
  ./lib/go/src
  ./services/service-one
  ./services/service-two
)
```

lib/go/src/go.mod:
```
module company.com/common
```

services/service-one/go.mod:
```
module company.com/service-one
```

Each service can then `import "company.com/common/grpc"` and everything resolves nicely. Building an application is simple too: `go build company.com/service-one/app`. We use `go work sync` to keep all the package dependencies consistent across services.

This all works very nicely. The only slight issue is that `go mod tidy` within a given service gives errors:

```
go: company.com/service-one/app imports
	company.com/common: cannot find module providing package company.com/common: unrecognized import path "company.com/common": https fetch: Get "https://company.com/common?go-get=1": dial tcp: lookup company.com: no such host
```

this is fine, because there's nothing to retrieve, but it would be ideal if it paid attention to the `use` directives of `go.work` and didn't bother to do the lookups.

**Comment by majelbstoat** [comment]
As for why the services aren't all in one module together - it's because they _aren't_ all one module. They are applications that happen to be written in go, amongst others that are not. They are each deployed separately. They have different environment variables, they have different kubernetes configs. We choose to organize our code by logical service, not by implementation language. 

This is service one:
```
├── app
│   ├── main.go
├── env
│   ├── dev.env
│   ├── production.env
│   └── test.env
├── go.mod
├── go.sum
└── k8s
    ├── dev.yaml
    ├── production.yaml
    └── test.yaml
```

This is service three:
```
├── app
│   ├── index.ts
├── env
│   ├── dev.env
│   ├── production.env
│   └── test.env
├── package.json
└── k8s
    ├── dev.yaml
    ├── production.yaml
    └── test.yaml
```

The enforced regularity of this layout greatly simplifies devops in a heterogeneous repo, and reduces cognitive load on developers as the system and team grows..

**Comment by matloob** [comment]
@majelbstoat I'd like to understand this better. Is the goal of placing the go.mod files in each of the service directories (rather than the top level) to make the versions of the dependencies of the services independent of each other?

**Comment by majelbstoat** [comment]
The idea is that each go.mod represents the minimal set of dependencies for each service/library. This helps optimize deployments, where we cache the go build cache and results of `go get` for each service in a cache keyed by the `go.sum` of that service (and the common library go.mod).

.circleci.yaml:
```
restore_cache:
      keys:
          - go-mod-v3-{{ checksum "lib/go/src/go.sum" }}-{{ checksum "services/<< parameters.service-name >>/go.sum"}}
```

Practically speaking, that means if we make a change to the dependencies of service-one, we only have to expire the cache for service-one, not every go service. If they were all in one `go.mod`, we would have to redeploy every service, and blow every cache, for every dependency change to every service, even if the other services didn't depend on the new dependency. (Or else make our deployment decision-making process quite a bit more complex.)

We actually want common dependencies for each service to be at the same version. For example, it's important that the version of the `google.golang.org/grpc` package is the same for all services. `go work sync` does this for us - it ensures all `go.mod`s that include `google.golang.org/grpc` include it at the same highest version. It gets run automatically pre-commit if any go.mod has changed.

Aesthetically too, each service imports packages from the common library. If they all were in the same go.mod, imports for a service from the common library would be like `import "company.com/lib/go/src/common/grpc"`, which kinda ugly and a pain to change if we ever changed its relative location in the repo. It doesn't happen often, but it has happened. In the current setup, it requires one edit to one `use` line in `go.work`.

**Comment by fayep** [comment]
I think that I have a simpler reason for why this is incorrect behavior right now.
go.mod and go.sum are distributable artifacts. go.work is for local overriding of dependency modules for more streamlined development.  Once development is done, I should not need to modify go.mod or go.sum as I will have pushed my simultaneously developed dependency module as well as this one.

Expected behavior:
I have an as yet unpublished dependency which I have added to my workspace `../mymod`.  If I `go get` or `go mod tidy` it should not attempt to locate `mymod` upstream because it is defined in my workspace.  It should continue to get other dependency modules not included in my workspace.

Observed behavior:
`go mod tidy` and `go get` retrieve dependency modules but fail on trying to load my unpublished dependency.  The newly retrieved modules are not added to `go.mod` and `go.sum` leaving the external dependencies unresolved.

Workaround:
add `replace myrepo.myid.mymod => ../mymod` which we used to do before workspaces.  This is unsatisfactory because once development is done, you need to go back into `go.mod` and remove this override.

**Comment by mytechcn** [comment]
When can this problem be solved

**Comment by matloob** [comment]
Thanks for the comments. We're still thinking about this. Let me know if you're interested in doing a video call to talk about your use case.

It still seems to me that changing the behavior of `go mod tidy` isn't correct. In the common case with a workspace that multiple modules are participating in, the module isn't tied to a specific workspace. So tidy should be able to tidy the module without knowledge of the workspace. Similarly, `go get` by default should operate on the module alone without knowledge of the workspace.

@fayep, It seems to me that a temporary `replace` directive is appropriate to use in this case. If the module doesn't exist upstream yet, you can't have a proper `require`ment on that module. The `replace` directive allows the module's build graph to be complete until that dependency module is released and can be required.

**Comment by mytechcn** [comment]
> Thanks for the comments. We're still thinking about this. Let me know if you're interested in doing a call to talk about your use case.
> 
> It still seems to me that changing the behavior of `go mod tidy` isn't correct. In the common case with a workspace that multiple modules are participating in, the module isn't tied to a specific workspace. So tidy should be able to tidy the module without knowledge of the workspace. Similarly, `go get` by default should operate on the module alone without knowledge of the workspace. So we'll need to think of other ideas to address these use cases.
> 
> [@fayep](https://github.com/fayep), It seems to me that a temporary `replace` directive is appropriate to use in this case. If the module doesn't exist upstream yet, you can't have a proper `require`ment on that module. The `replace` directive allows the module's build graph to be complete until that dependency module is released and can be required. This is especially important to tidy a module. It would be nice if we could reduce the friction in the go get use case though.

--------------------------------------------------------------------------
I have a common module that includes some common functions, which are referenced by all projects. For example, project A now references it in the workspace, which is private.
 But now when A uses go mod tidy to manage project dependencies, it will report an error.
It is necessary for multiple modules to reference each other, which is a common basic feature in Java and C #. This bug should undoubtedly be resolved as soon as possible.
 We have been dragging on for too long.

**Comment by spekary** [comment]
Since there are multiple situations where you may or may not want the behavior, why not add a flag for go mod tidy to pay attention to go.work?

**Comment by matloob** [comment]
@spekary As we currently define it, a tidy module is that it is one that includes the appropriate set of requirements to satisfy its imports. Changing or adding a concept of being tidy with respect to a workspace would complicate things further and modules can be fairly complex as is. We'd like to find a solution that wouldn't add to the complexity of the system.

In many of the cases listed on this issue the workspaces do logically function as a single module (for example they are all versioned together, and have the same set of dependency versions) but there are various blockers or concerns that prevent the use of a single module. And there are cases where the modules do function differently but there's friction in the process of making them available to the modules that depend on them via `go get` We'd like to address those issues for the users facing them. And for those users facing different issues, we'd like to talk to them about their use cases to try to better understand the problem space.

**Comment by spekary** [comment]
I suppose I am suggesting approaching the problem differently. When a developer adds a go.work file, they are saying that they are doing something special. I thought that was the idiom. They are composing a working environment that anticipates eventually removing the go.work file, running go mod tidy, and checking in the go.mod file. But for the time being, they need an environment that can work differently. 

Before go.work, we would put "replace" statements into the go.mod file, but then we would have to remember not to check that in. go.work was supposed to solve that problem, but I guess it didn't, because go.mod still would need to behave differently because of the dependencies implied in the use of different packages, and so we still have to remember not to check in go.mod during development. I guess we need a go.mod.work file (and of course the ensuing go.mod.sum.work file).

So how about this: When go.work is present, go mod tidy creates a go.mod.work file. When a go.mod.work file is present, the various go tools use it instead of go.mod. Yeah, it could create some confusion if the developer doesn't realize that is what is happening. Perhaps the go tools could call that out while processing files, like pointing out that a go.mod.work file was found? In any case, the developer started the process by creating a go.work file, and the developer can fix the problem by removing it and the go.mod.work file.

The alternative I was suggesting previously is to use a flag to cause go mod tidy to behave this way. But as I think about it, that flag would have to be across all go tools to tell them to use go.mod.work if it exists. Not an easy problem to solve, but I prefer Go's general use of convention to solve these problems, rather than a proliferation of flags in the tools.

**Comment by matloob** [comment]
> I suppose I am suggesting approaching the problem differently. When a developer adds a go.work file, they are saying that they are doing something special. I thought that was the idiom. They are composing a working environment that anticipates eventually removing the go.work file, running go mod tidy, and checking in the go.mod file. But for the time being, they need an environment that can work differently.

Yes, a `go.work` is saying that the user is doing something special.

> Before go.work, we would put "replace" statements into the go.mod file, but then we would have to remember not to check that in. go.work was supposed to solve that problem, but I guess it didn't, because go.mod still would need to behave differently because of the dependencies implied in the use of different packages, and so we still have to remember not to check in go.mod during development. I guess we need a go.mod.work file (and of course the ensuing go.mod.sum.work file).

Are you seeing issues in the case where the dependency module already exists? If you are I'd like to hear more about the specifics. There is definitely more friction in the case where the dependency module doesn't yet exist. We may have to think more about those cases but they should only happen "once" when the dependency module is being introduced.

> So how about this: When go.work is present, go mod tidy creates a go.mod.work file. When a go.mod.work file is present, the various go tools use it instead of go.mod. Yeah, it could create some confusion if the developer doesn't realize that is what is happening. Perhaps the go tools could call that out while processing files, like pointing out that a go.mod.work file was found? In any case, the developer started the process by creating a go.work file, and the developer can fix the problem by removing it and the go.mod.work file.
> 
> The alternative I was suggesting previously is to use a flag to cause go mod tidy to behave this way. But as I think about it, that flag would have to be across all go tools to tell them to use go.mod.work if it exists. Not an easy problem to solve, but I prefer Go's general use of convention to solve these problems, rather than a proliferation of flags in the tools.

I'd like to hear more about your use case. In the case where you're using `go.work` to do something special (and the dependency module already exists at a previous version in a `go get`table state) you should be able to work on both modules, go getting dependencies as needed. You may not be able to tidy until you release the dependency module if the changes made compared to the previous version affect the dependency graph, but tidying the module shouldn't be necessary until you release. Are you seeing issues running `go get`?

**Comment by matloob** [comment]
@nemith Yeah the use case you mentioned where you're adding a new module that hasn't been released yet has some friction currently. Just to make sure I understand your use case, you weren't trying to do module-mode operations like `go get` of a dependency or `go mod tidy` but instead couldn't even do a `go build` in workspace mode, right?

In that case, currently we'd recommend adding a `replace` directive until there's a module version that could be properly `require`d. Though I can see how it might get confusing to have to use a `replace` together with a workspace. We definitely need to improve our documentation around that. And try to see if there are ways to make that use case better.

**Comment by nemith** [comment]
@matloob The most friction was when adding a new module with inter-mod dependencies.  A replace could have fixed it but additional additional load of updating docker files to include sources for dependent modules,  using a path-filter with GHA to fix CI on dependent changes, on top of adding `replace` directives everywhere just wasn't worth it.  

Being naive and not knowing better i tried `go.work` as a way to make modules "aware" of each other, the obviously isn't the case.  Also I am not even sure what the valid use cases for go.work is anymore if `replace` is still a valid and promoted  option.  Confusion is the right word.

It's sad as a monorepo approach solves a lot of other issues of dependency testing and integration.

**Comment by ohir** [comment]
@matloob 
> adding a new module that hasn't been released yet has some friction currently.

It is not a "friction" - for the bootstrap and air-gapped environments it is entirely counterproductive.  Even in open source scenarios it makes the bootstraping phase a pain.  Frankly, I can not see where and how _workspaces_ current functionality might give less work.  I personally am back to the _mod_ replaces managed by a script.  Making a workspace just muds the water, makes project brittle, and doubles the maintenance time.

**Comment by crazy-matt** [comment]
@matloob If you care about downstream consumers, then don't forget about just all the other devs who do not see why go workspaces can't address that issue and go mod tidy handle it if one forces it to do so. Bootstrapping and CI/CD are valid use cases where it's currently painful with replace directives management.
So why not just adding a flag to `go mod tidy` to make it workspace aware? That way, the current implementation is still valid and not breaking any downstream consumer implementation. If one wants to use that flag then assume they know what they're doing, if they don't, they will learn quick. This issue has been raise 3 years ago.

**Comment by matloob** [comment]
We've spent some time thinking about how to improve the experience when working with workspace modules that haven't been released yet. For Go 1.25 we plan to improve `go get` to work in modules in workspaces that have unreleased modules. We're going to close this issue because we do not plan to change the behavior of `go mod tidy`. A module can't be tidy unless it requires all of its dependencies. We recommend users run `go get` to find and add requirements for missing dependencies, and then `go work sync` to sync and clean up workspace modules.

I've filed issue #73654 to track the work to fix `go get`.

If  `go work sync` is producing errors in workspace modules, or it is trying to resolve a module that's in the workspace, that's a bug. I filed issue #73655 to track that.

I think these should address many of the use cases in workspace mode that tidy is used for in module mode: namely finding modules to satisfy missing imports, and cleaning up the requirements in modules.

We may want to think about whether it makes sense to extend `go work sync` to have an option to run `go get` in each of the workspace's modules before doing the sync but that would require more thought As an example of an issue, that wouldn't work if more multiple workspace modules were missing imports that would need to be satisfied by different modules.
